File: Problems/0395-longest-substring-with-at-least-k-repeating-characters.py
class Solution(object):
    def longestSubstring(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        ans = 0

        for target in range(1, 27):
            cnt = [0] * 26
            left = right = 0
            unique = at_least_k = 0

            while right < len(s):
                idx = ord(s[right]) - 97
                if cnt[idx] == 0:
                    unique += 1
                cnt[idx] += 1
                if cnt[idx] == k:
                    at_least_k += 1
                right += 1

                while unique > target:
                    idx = ord(s[left]) - 97
                    if cnt[idx] == k:
                        at_least_k -= 1
                    cnt[idx] -= 1
                    if cnt[idx] == 0:
                        unique -= 1
                    left += 1

                if unique == target and unique == at_least_k:
                    ans = max(ans, right - left)

        return ans

File: Problems/0395-longest-substring-with-at-least-k-repeating-characters.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here

class Solution {
public:
    int longestSubstring(string s, int k) {
        int ans = 0, n = s.size();

        for (int target = 1; target <= 26; target++) {
            int cnt[26] = {};
            int left = 0, right = 0;
            int unique = 0, atLeastK = 0;

            while (right < n) {
                int idx = s[right] - 'a';

                if (cnt[idx] == 0)
                    unique++;

                cnt[idx]++;

                if (cnt[idx] == k)
                    atLeastK++;

                right++;

                while (unique > target) {
                    idx = s[left] - 'a';

                    if (cnt[idx] == k)
                        atLeastK--;

                    cnt[idx]--;

                    if (cnt[idx] == 0)
                        unique--;

                    left++;
                }

                if (unique == target && unique == atLeastK)
                    ans = max(ans, right - left);
            }
        }

        return ans;
    }
};
